drivers/flash/spi_flash_at45d.c

```diff
@@ -0,0 +1,692 @@
+		}
+	};
+	const struct spi_buf rx_buf[] = {
+		{
+			.len = sizeof(opcode),
+		},
+		{
+			.buf = read_id,
+			.len = sizeof(read_id),
+		}
+	};
+	DEF_BUF_SET(tx_buf_set, tx_buf);
+	DEF_BUF_SET(rx_buf_set, rx_buf);
+
+	err = spi_transceive(get_dev_data(dev)->spi,
+			     &cfg->spi_cfg,
+			     &tx_buf_set, &rx_buf_set);
+	if (err != 0) {
+		LOG_ERR("SPI transaction failed with code: %d/%u",
+			err, __LINE__);
+		return -EIO;
+	}
+
+	if (memcmp(expected_id, read_id, sizeof(read_id)) != 0) {
+		LOG_ERR("Wrong JEDEC ID: %02X %02X %02X, "
+			"expected: %02X %02X %02X",
+			read_id[0], read_id[1], read_id[2],
+			expected_id[0], expected_id[1], expected_id[2]);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+/*
+ * Reads 2-byte Status Register:
+ * - Byte 0 to LSB
+ * - Byte 1 to MSB
+ * of the pointed parameter.
+ */
+static int read_status_register(struct device *dev, u16_t *status)
+{
+	int err;
+	const u8_t opcode = CMD_READ_STATUS;
+	const struct spi_buf tx_buf[] = {
+		{
+			.buf = (void *)&opcode,
+			.len = sizeof(opcode),
+		}
+	};
+	const struct spi_buf rx_buf[] = {
+		{
+			.len = sizeof(opcode),
+		},
+		{
+			.buf = status,
+			.len = sizeof(u16_t),
+		}
+	};
+	DEF_BUF_SET(tx_buf_set, tx_buf);
+	DEF_BUF_SET(rx_buf_set, rx_buf);
+
+	err = spi_transceive(get_dev_data(dev)->spi,
+			     &get_dev_config(dev)->spi_cfg,
+			     &tx_buf_set, &rx_buf_set);
+	if (err != 0) {
+		LOG_ERR("SPI transaction failed with code: %d/%u",
+			err, __LINE__);
+	}
+
+	return (err != 0) ? -EIO : 0;
+}
+
+static int wait_until_ready(struct device *dev)
+{
+	int err;
+	u16_t status;
+
+	do {
+		err = read_status_register(dev, &status);
+	} while (err == 0 && !(status & 0x80));
+
+	return err;
+}
+
+static int configure_page_size(struct device *dev)
+{
+	int err;
+	u16_t status;
+	u8_t const conf_binary_page_size[] = { CMD_BINARY_PAGE_SIZE };
+	const struct spi_buf tx_buf[] = {
+		{
+			.buf = (void *)conf_binary_page_size,
+			.len = sizeof(conf_binary_page_size),
+		}
+	};
+	DEF_BUF_SET(tx_buf_set, tx_buf);
+
+	err = read_status_register(dev, &status);
+	if (err != 0) {
+		return err;
+	}
+
+	/* If the device is already configured for "power of 2" binary
+	 * page size, there is nothing more to do.
+	 */
+	if (status & 0x01) {
+		return 0;
+	}
+
+	err = spi_write(get_dev_data(dev)->spi,
+			&get_dev_config(dev)->spi_cfg,
+			&tx_buf_set);
+	if (err != 0) {
+		LOG_ERR("SPI transaction failed with code: %d/%u",
+			err, __LINE__);
+	} else {
+		err = wait_until_ready(dev);
+	}
+
+	return (err != 0) ? -EIO : 0;
+}
+
+static bool is_valid_request(off_t addr, size_t size, size_t chip_size)
+{
+	return (addr >= 0 && (addr + size) <= chip_size);
+}
+
+static int spi_flash_at45d_read(struct device *dev, off_t offset,
+				void *data, size_t len)
+{
+	const struct spi_flash_at45d_config *cfg = get_dev_config(dev);
+	int err;
+
+	if (!is_valid_request(offset, len, cfg->chip_size)) {
+		return -ENODEV;
+	}
+
+	u8_t const op_and_addr[] = {
+		[0] = CMD_READ,
```

Comment:
I don't think designated initializers add any value when it's a contiguous gapless sequence using literal ordinals.  A standard initializer list would be less confusing.